core/rfb.js

```diff
@@ -482,60 +456,66 @@
                 this._sock.off('close');  // make sure we don't get a double event
             }
 
+            this._onUpdateState(this, state, oldstate);
             switch (state) {
+                case 'disconnected':
+                    this._cleanupSocket(); // Make sure we are closed.
+
+                    if (typeof(errorMsg) !== 'undefined') {
+                        this._onDisconnected(this, errorMsg);
+                    } else {
+                        this._onDisconnected(this); // No msg means clean disconnect
+                    }
+                    break;
+
                 case 'connected':
-                    if (oldstate === 'disconnected' || oldstate === 'failed') {
-                        Util.Error("Invalid transition from 'disconnected' or 'failed' to 'connected'");
+                    if (oldstate === 'disconnected') {
+                        Util.Error("Invalid transition from 'disconnected' to 'connected'");
                     }
                     break;
 
                 case 'connecting':
+                    // A previous connect may asynchronously cause a connection
+                    // so make sure we are still closed.
+                    this._cleanupSocket();
```

Comment:
There is one issue remaining here with this approach, and that's the fact that the RFB object leaves stray event listeners around. These need to be cleaned up on disconnect so that all references are properly dropped. All the ones left over originate in devices.js.